docker/generate_makefile.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simple generator for local Makefile rules."""
+
+import os
+import yaml
+
+BASE_TAG = 'gcr.io/fuzzbench'
+BENCHMARKS_DIR = os.path.join(os.path.dirname(__file__), os.pardir,
+                              'benchmarks')
+FUZZERS_DIR = os.path.join(os.path.dirname(__file__), os.pardir, 'fuzzers')
+
+BOILERPLATE = """
+cache_from = $(if ${RUNNING_ON_CI},--cache-from {fuzzer},)
+"""
+
+FUZZER_TEMPLATE = """
+.{fuzzer}-builder: base-builder
+	docker build \\
+    --tag {base_tag}/builders/{fuzzer} \\
+    --file fuzzers/{fuzzer}/builder.Dockerfile \\
+    $(call cache_from,{base_tag}/builders/{fuzzer}) \\
+    fuzzers/{fuzzer}
+
+.pull-{fuzzer}-builder: pull-base-builder
+	docker pull {base_tag}/builders/{fuzzer}
+"""
+
+FUZZER_BENCHMARK_TEMPLATE = """
+.{fuzzer}-{benchmark}-builder: .{fuzzer}-builder
+	docker build \\
+    --tag {base_tag}/builders/{fuzzer}/{benchmark} \\
+    --build-arg fuzzer={fuzzer} \\
+    --build-arg benchmark={benchmark} \\
+    $(call cache_from,{base_tag}/builders/{fuzzer}/{benchmark}) \\
+    --file docker/benchmark-builder/Dockerfile \\
+    {build_arguments} \\
+    .
+
+.pull-{fuzzer}-{benchmark}-builder: .pull-{fuzzer}-builder
+	docker pull {base_tag}/builders/{fuzzer}/{benchmark}
+
+ifneq ({fuzzer}, coverage)
```

Comment:
Consider doing logic like this in the generation script itself. I believe that'd be easier to read.

Reply:
I'm not opposed to this but I'm honestly not sure that it's cleaner. If this is removed we need several subtemplates which also have to be formatted ahead of time and combined together, which feels complicated.

Any thoughts?

Reply:
You might consider having one template per docker image type for this, but of course if it just further complicates things then never mind.

Reply:
It would complicate updates to the portions outside of the if/else as they'd need to be updated twice. I think I'll leave it as-is, but if there would ever be cases where more things are handled conditionally I think it would make sense to refactor this to actually include the logic in python rather than make.

Reply:
Sounds good thanks

Reply:
I prefer breaking up the logic so that we could have one source of truth for GCB yaml files and make. But I'm with landing this for now. Probably better to figure out what that logic looks like when we generate the yaml files as well, rather than ahead of time.

Reply:
+1, landing this.